File: tests/test_wire_contracts.mjs
import test from "node:test";
import assert from "node:assert/strict";
import fs from "node:fs";
import { WIRE_VERSION, PROVIDERS, JOB_STATUSES, JOB_TRANSITIONS, THINKING_MODES,
  normalizeProviderSettings, normalizeSnapshotWire, normalizeJobWire, assertJobTransition, assertObservedJobTransition } from "../web/js/prompt-studio/core/wire-contracts.js";
import { LLM_THINKING_MODE_OPTIONS, LLM_PROFILE_STORAGE_KEY } from "../web/js/prompt-studio/core/constants.js";
import { normalizeLlmProfile, loadLlmProfiles } from "../web/js/prompt-studio/settings/llm-profile-store.js";
import { normalizeGenerationSnapshot } from "../web/js/prompt-studio/chat/generation-state.js";

const read = name => JSON.parse(fs.readFileSync(new URL(`../web/js/prompt-studio/core/${name}.json`, import.meta.url), "utf8"));
const fixtures = read("wire-contract-fixtures");

test("shared JS enums and legacy profile options match the canonical Python source", () => {
  const enums = read("wire-contract-enums");
  assert.equal(WIRE_VERSION, enums.wire_version);
  assert.deepEqual(PROVIDERS, enums.providers);
  assert.deepEqual(JOB_STATUSES, enums.job_statuses);
  assert.deepEqual(JOB_TRANSITIONS, enums.job_transitions);
  assert.deepEqual(THINKING_MODES, enums.thinking_modes);
  assert.deepEqual(LLM_THINKING_MODE_OPTIONS, enums.thinking_modes);
});

for (const kind of ["provider", "job", "snapshot"]) test(`${kind} persisted legacy fixtures migrate explicitly`, () => {
  const normalize = { provider: normalizeProviderSettings, job: normalizeJobWire, snapshot: normalizeSnapshotWire }[kind];
  for (const { input, expected } of fixtures[`${kind}_legacy`]) {
    const original = structuredClone(input);
    const actual = normalize(input);
    if (kind === "provider") for (const [key, value] of Object.entries(expected)) assert.deepEqual(actual[key], value);
    else assert.deepEqual(actual, expected);
    assert.deepEqual(input, original);
  }
});

test("versioned provider discriminants, misspelled fields and invalid numbers are rejected", () => {
  assert.throws(() => normalizeProviderSettings({ wire_version: 2 }), /version/);
  assert.throws(() => normalizeProviderSettings({ wire_version: 1, llm_provider: "typo" }), /llm_provider/);
  assert.throws(() => normalizeProviderSettings({ llm_provder: "ollama" }, { strict: true }), /Unknown/);
  for (const value of [NaN, Infinity, null, 6]) assert.throws(() => normalizeProviderSettings({ temperature: value }), /temperature/);
});

test("new snapshots require both envelopes and preserve unknown metadata", () => {
  for (const snapshot of [{ output: {} }, { workflow: {} }, { workflow: [], output: {} }]) {
    assert.throws(() => normalizeSnapshotWire({ wire_version: 1, kind: "workflow_snapshot", snapshot }));
  }
  const source = fixtures.snapshot_legacy[0].input;
  const migrated = normalizeSnapshotWire(source);
  migrated.snapshot.custom.keep = false;
  assert.equal(source.custom.keep, true);
});

test("Image saved generation snapshots retain workflow and all metadata; output-only is explicit legacy", () => {
  const source = fixtures.snapshot_legacy[0].input;
  const normalized = normalizeGenerationSnapshot(source);
  assert.deepEqual(normalized, source);
  normalized.workflow.nodes.push({ id: 1 });
  assert.equal(source.workflow.nodes.length, 0);
  assert.deepEqual(normalizeGenerationSnapshot({ output: { one: {} }, legacyMetadata: "keep" }), { output: { one: {} }, legacyMetadata: "keep" });
  assert.equal(normalizeGenerationSnapshot({ output: {}, workflow: [] }), null);
  assert.equal(normalizeGenerationSnapshot({ workflow: {} }), null);
});

test("all job transitions follow the published graph and terminal payloads are required", () => {
  for (const from of JOB_STATUSES) for (const to of JOB_STATUSES) {
    if (JOB_TRANSITIONS[from].includes(to)) assert.doesNotThrow(() => assertJobTransition(from, to));
    else assert.throws(() => assertJobTransition(from, to), /transition/);
  }
  assert.throws(() => normalizeJobWire({ job_id: "job", status: "complete" }), /result/);
  assert.throws(() => normalizeJobWire({ job_id: "job", status: "failed" }), /error/);
  assert.doesNotThrow(() => assertObservedJobTransition("queued", "complete"));
  assert.doesNotThrow(() => assertObservedJobTransition("running", "running"));
  assert.throws(() => assertObservedJobTransition("running", "queued"), /transition/);
  assert.throws(() => assertObservedJobTransition("failed", "running"), /transition/);
});

test("old profile arrays and Qwen presets migrate through the shared Video/Image reader", () => {
  const values = new Map([[LLM_PROFILE_STORAGE_KEY, JSON.stringify([{ id: "qwen3.5", name: "Qwen3.5", temperature: 0.8 }])]]);
  globalThis.localStorage = { getItem: key => values.get(key) ?? null, setItem: (key, value) => values.set(key, value) };
  try {
    const profiles = loadLlmProfiles();
    assert.equal(profiles.find(profile => profile.id === "qwen3.5").name, "Default");
    assert.ok(profiles.some(profile => profile.id === "qwen3.8-27b"));
    assert.equal(JSON.parse(values.get(LLM_PROFILE_STORAGE_KEY)).version, 6);
    assert.equal(normalizeLlmProfile({ thinking_modes: ["invalid"] }).thinking_mode, "Disabled");
  } finally { delete globalThis.localStorage; }
});

File: tests/test_wire_contracts.py
import json
from pathlib import Path
import unittest
from typing import get_args

import wire_contracts as contracts


class WireContractsTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.fixtures = json.loads((Path(__file__).resolve().parents[1] / "web/js/prompt-studio/core/wire-contract-fixtures.json").read_text(encoding="utf-8"))

    def test_typed_enums_match_canonical_data(self):
        self.assertEqual(get_args(contracts.LlmProvider), contracts.PROVIDERS)
        self.assertEqual(get_args(contracts.JobStatus), contracts.JOB_STATUSES)

    def test_legacy_fixtures(self):
        for kind, normalize in (("provider", contracts.normalize_provider_settings), ("job", contracts.normalize_job_wire), ("snapshot", contracts.normalize_snapshot_wire)):
            for fixture in self.fixtures[kind + "_legacy"]:
                with self.subTest(kind=kind, fixture=fixture):
                    before = json.dumps(fixture["input"], sort_keys=True)
                    result = normalize(fixture["input"])
                    if kind == "provider":
                        for key, expected in fixture["expected"].items():
                            self.assertEqual(result[key], expected)
                    else:
                        self.assertEqual(result, fixture["expected"])
                    self.assertEqual(json.dumps(fixture["input"], sort_keys=True), before)

    def test_provider_version_enum_unknown_fields_and_bad_numbers(self):
        for data in ({"wire_version": 2}, {"wire_version": True}, {"wire_version": 1, "llm_provider": "typo"}, {"temperature": None}, {"temperature": float("nan")}, {"temperature": 6}):
            with self.subTest(data=data), self.assertRaises(ValueError):
                contracts.normalize_provider_settings(data)
        with self.assertRaisesRegex(ValueError, "Unknown provider fields"):
            contracts.normalize_provider_settings({"llm_provder": "ollama"}, strict=True)

    def test_new_snapshot_envelope_is_required_and_copy_is_immutable(self):
        for snapshot in ({"workflow": {}}, {"output": {}}, {"workflow": [], "output": {}}):
            with self.subTest(snapshot=snapshot), self.assertRaises(ValueError):
                contracts.normalize_snapshot_wire({"wire_version": 1, "kind": "workflow_snapshot", "snapshot": snapshot})
        original = self.fixtures["snapshot_legacy"][0]["input"]
        result = contracts.normalize_snapshot_wire(original)
        result["snapshot"]["custom"]["keep"] = False
        self.assertTrue(original["custom"]["keep"])

    def test_all_job_transitions_and_terminal_payloads(self):
        for previous in contracts.JOB_STATUSES:
            for following in contracts.JOB_STATUSES:
                if following in contracts.JOB_TRANSITIONS[previous]:
                    contracts.assert_job_transition(previous, following)
                else:
                    with self.assertRaises(ValueError):
                        contracts.assert_job_transition(previous, following)
        for status in ("complete", "failed", "misspelled"):
            with self.assertRaises(ValueError):
                contracts.normalize_job_wire({"job_id": "job", "status": status})
